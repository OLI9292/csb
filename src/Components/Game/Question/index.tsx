import React from "react"
import { Animated } from "react-native"

import Prompt from "./prompt"
import Answer, { UserAnswer } from "./answer"
import Choices from "./choices"
import _ from "underscore"
import { getUser, logQuestionHistory } from "../../../Models/user"
import { logQuestions } from "../../../Models/question"
import { secondsDiff } from "../../../lib/helpers"

export interface Props {
  question: any
  questionDone: () => void
  isInterlude: boolean
  questionSequenceEnded: boolean
}

export interface QuestionLog {
  id: string
  questionCategory: string
  questionSubCategory: string
  questionIdentifier: string
  questionPrompt: string
  questionSecondaryPrompt: string
  startTime: Date
  endTime?: Date
  secondsTaken?: number
  guesses: string
  userId: string
  email: string
  correct: boolean
  finished: boolean
}

interface State {
  question?: any
  user?: any
  startTime?: Date
  opacityAnimation: Animated.Value
  userAnswers: string[]
  choices: any[][]
  currentChoices: string[]
  choiceTree?: any
  questionsLog: QuestionLog[]
  isCorrecting: boolean
}

export default class Question extends React.Component<Props, State> {
  private choicesComponent: React.RefObject<Choices>
  private promptComponent: React.RefObject<Prompt>
  private nextQuestionTimeout: any

  constructor(props: Props) {
    super(props)

    this.state = {
      isCorrecting: false,
      currentChoices: [],
      choices: [],
      userAnswers: [],
      questionsLog: [],
      opacityAnimation: new Animated.Value(1),
    }

    this.promptComponent = React.createRef()
    this.choicesComponent = React.createRef()
  }

  async componentDidMount() {
    const user = await getUser() // TODO: - implement better state mgmt and remove
    this.setState({ user })
    this.reset(this.props.question)
  }

  componentWillUnmount() {
    const questionsLog = this.props.questionSequenceEnded
      ? (this.updateRecord("finished", undefined, true) as QuestionLog[])
      : this.state.questionsLog

    if (questionsLog.length === 0) {
      return
    }

    // Save in question 2 table
    logQuestions(questionsLog)

    // Save under user model
    const _id = this.state.user._id
    const question2History = questionsLog.filter(q => q.finished).map(q => ({ id: q.id, perfect: q.correct }))
    logQuestionHistory(JSON.stringify({ _id, question2History }))
  }

  componentWillReceiveProps(nextProps: Props) {
    if (nextProps.question && this.props.question !== nextProps.question) {
      Animated.timing(this.state.opacityAnimation, { toValue: 0, duration: 500 }).start()

      this.updateRecord("finished")

      this.nextQuestionTimeout = setTimeout(() => {
        this.reset(nextProps.question)
        Animated.timing(this.state.opacityAnimation, { toValue: 1, duration: 500 }).start()
      }, 500)
    }
  }

  cleanChoiceTreeJson(json: any) {
    const choiceTree = JSON.parse(json)

    const clean = (obj: any): any => {
      if (obj.branches) {
        return clean(obj.branches)
      } else if (_.isEmpty(obj) || _.isEqual(_.keys(obj), ["choiceSetId"])) {
        return undefined
      } else {
        return _.mapObject(obj, (v, k) => clean(v))
      }
    }

    const answers = choiceTree.answers.slice(1)
    const tree = clean(choiceTree.tree)
    return { answers, tree }
  }

  record(question: any) {
    const { user, questionsLog } = this.state

    const mapText = (prompt: any[]) =>
      prompt.map(c => (c.highlight ? `***${c.value as string}***` : (c.value as string))).join(" ")

    const questionLog: QuestionLog = {
      id: question._id,
      questionCategory: question.category,
      questionSubCategory: question.subCategory,
      questionIdentifier: question.identifier,
      questionPrompt: mapText(question.prompt),
      questionSecondaryPrompt: mapText(question.secondaryPrompt),
      startTime: new Date(),
      guesses: "",
      userId: user._id,
      email: user.email,
      correct: true,
      finished: false,
    }

    questionsLog.push(questionLog)
    this.setState({ questionsLog })
  }

  updateRecord(attr: string, value?: string, dismounting: boolean = false): QuestionLog[] | undefined {
    const questionsLog = this.state.questionsLog
    const log = _.last(questionsLog) as QuestionLog

    if (attr === "guess") {
      log.guesses.length && (log.guesses += "; ")
      log.guesses += value
    } else if (attr === "correct") {
      log.correct = false
    } else if (attr === "finished") {
      log.finished = true
      log.endTime = new Date()
      log.secondsTaken = secondsDiff(log.startTime, log.endTime)
    }

    if (dismounting) {
      return questionsLog
    } else {
      this.setState({ questionsLog })
    }
  }

  reset(question: any) {
    const choiceTree = question.choiceTreeJson && this.cleanChoiceTreeJson(question.choiceTreeJson)
    this.record(question)

    this.setState(
      {
        startTime: new Date(),
        choiceTree: choiceTree,
        choices: question.choices,
        question: question,
        isCorrecting: false,
        userAnswers: [],
      },
      this.checkInput
    )
  }

  checkInput() {
    let { choiceTree, choices, userAnswers } = this.state

    if (choiceTree) {
      let { answers, tree } = choiceTree

      while (userAnswers.length) {
        const stop = this.state.isCorrecting && userAnswers[0] !== answers[0]
        if (stop) {
          break
        }
        tree = tree[userAnswers[0]]
        userAnswers = userAnswers.slice(1)
        answers = answers.slice(1)
      }

      if (tree) {
        const currentChoices = _.keys(tree)
        this.setState({ currentChoices })
      } else {
        this.setState(
          { isCorrecting: true },
          () => (this.questionDone() ? this.props.questionDone() : this.checkInput())
        )
      }
    } else if (choices) {
      const layer = choices[this.state.isCorrecting ? this.incorrectIndex(false) : userAnswers.length]

      if (layer) {
        const currentChoices = layer.map(c => c.value)
        this.setState({ currentChoices })
      } else {
        this.setState(
          { isCorrecting: true },
          () => (this.questionDone(false) ? this.props.questionDone() : this.checkInput())
        )
      }
    }
  }

  questionDone(isUsingChoiceTree: boolean = true): boolean {
    return this.incorrectIndex(isUsingChoiceTree) === -1
  }

  incorrectIndex(isUsingChoiceTree: boolean = true): number {
    const { userAnswers, choices, choiceTree } = this.state

    let index, slicedUserAnswers

    if (isUsingChoiceTree) {
      slicedUserAnswers = userAnswers.slice(0, choiceTree.answers.length)
      index = _.findIndex(
        _.zip(slicedUserAnswers, choiceTree.answers),
        (pair: string[]): boolean => pair[0] !== pair[1]
      )
    } else {
      index = _.findIndex(
        userAnswers,
        (a: string, idx: number) => !_.includes(choices[idx].filter(c => c.correct).map(c => c.value), a)
      )
    }

    if (index === -1 && isUsingChoiceTree) {
      this.setState({ userAnswers: slicedUserAnswers || userAnswers })
    } else {
      this.updateRecord("correct")
    }

    return index
  }

  guessed(value: string): void {
    let { choiceTree, choices, isCorrecting, userAnswers } = this.state
    this.updateRecord("guess", value)

    if (choiceTree) {
      if (isCorrecting) {
        const incorrectIndex = this.incorrectIndex()
        userAnswers[incorrectIndex] = value
        isCorrecting = incorrectIndex !== userAnswers.length - 1
      } else {
        userAnswers[userAnswers.length] = value
      }

      this.setState({ isCorrecting, userAnswers }, this.checkInput)
    } else if (choices) {
      userAnswers[isCorrecting ? this.incorrectIndex(false) : userAnswers.length] = value
      this.setState({ userAnswers }, this.checkInput)
    }
  }

  render() {
    const { userAnswers, question, choiceTree, choices, isCorrecting, currentChoices } = this.state

    if (!question) {
      return null
    }

    return (
      <Animated.View
        style={[
          {
            opacity: this.state.opacityAnimation,
            flex: 12,
            alignSelf: "stretch",
            marginLeft: 10,
            marginRight: 10,
          },
        ]}
      >
        <Prompt ref={this.promptComponent} prompt={question.prompt} secondaryPrompt={question.secondaryPrompt} />

        <Answer isCorrecting={isCorrecting} choiceTree={choiceTree} choices={choices} userAnswers={userAnswers} />

        <Choices
          isInterlude={this.props.isInterlude}
          ref={this.choicesComponent}
          data={currentChoices}
          guessed={this.guessed.bind(this)}
        />
      </Animated.View>
    )
  }
}
