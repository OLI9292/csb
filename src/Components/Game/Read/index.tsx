import * as React from "react"
import { Image } from "react-native"

import Markdown from "react-native-markdown-renderer"
import styled from "styled-components/native"

import Text from "../../Common/Text"
import FlexedView from "../../Common/FlexedView"
import Button from "../../Common/Button"
import { colors } from "../../../lib/colors"

export interface Read {
  value: string
  imgSrc?: string
}

interface Props {
  data: Read
  nextQuestion: () => void
}

export default class ReadComponent extends React.Component<Props, any> {
  public render() {
    const { imgSrc, value } = this.props.data

    const text = value.startsWith('{"markdown":') ? (
      <Markdown>{JSON.parse(value).markdown}</Markdown>
    ) : (
      <Text.m>{value}</Text.m>
    )

    return (
      <View>
        {imgSrc && (
          <FlexedView>
            <Image source={{ uri: `https://s3.amazonaws.com/${imgSrc}` }} />
          </FlexedView>
        )}

        {text}

        <FlexedView justifyContent="center">
          <Button color={colors.blue} text="Next" onPress={this.props.nextQuestion.bind(this)} />
        </FlexedView>
      </View>
    )
  }
}

const View = styled.View`
  display: flex;
  flex-direction: column;
  flex-grow: 1;
  justify-content: space-between;
  align-items: stretch;
  flex: 8;
  padding: 20px;
`
