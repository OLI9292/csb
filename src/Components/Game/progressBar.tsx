import React from 'react'
import styled from "styled-components/native"
import { colors, lighten10l } from '../../lib/colors'
import { Animated } from "react-native"

export interface Props {
  completion: number
}

interface State {
  animation: Animated.Value,
  width?: any
}

export default class ProgressBar extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props)
    this.state = {
      animation: new Animated.Value(0)
    }
  }

  componentWillReceiveProps(nextProps: Props) {    
    if (this.props.completion !== nextProps.completion) {
      this.animate(nextProps.completion)
    }
  }  

  animate(completion: number) {
    const toValue = (this.state.width || 0.5) * completion
    Animated.timing(this.state.animation, { toValue: toValue, duration: 500 }).start()
  }

  render() {
    return (
      <ContainerView>
        
        <Background
          onLayout={event => this.setState({ width: event.nativeEvent.layout.width })} />
        
        <Animated.View style={[{ width: this.state.animation }]}>
          <Progress />
        </Animated.View>
        
      </ContainerView>
    )
  }
}

const ContainerView = styled.View`
  flex: 8;
  justify-content: center;
  margin: 0px 25px;
`;

const Background = styled.View`
  height: 10px;
  width: 100%;
  border-radius: 5px;
  background-color: ${colors.lightGray};
`

const Progress = styled.View`
  height: 10px;
  margin-top: -10px;
  width: 100%;
  border-radius: 5px;
  background-color: ${colors.yellow};
`
