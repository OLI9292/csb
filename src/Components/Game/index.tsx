import React from 'react'
import styled from "styled-components/native"
import Icon from 'react-native-vector-icons/MaterialCommunityIcons';

import { colors, lighten10l } from '../../lib/colors'
import Question from './Question/index'
import ProgressBar from './progressBar'

export interface Props {
  questions: any[],
  navigator: any
}

interface State {
  questionIndex: number,
  questionDone: boolean
}

export default class Game extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props)
    this.state = {
      questionIndex: 0,
      questionDone: false
    }
  }

  questionDone() {
    this.setState({ questionDone: true })
  }

  nextQuestion() {
    const questionIndex = this.state.questionIndex + 1
    
    this.setState({
      questionIndex: questionIndex,
      questionDone: false
    })

    if (questionIndex === this.props.questions.length) {
      setTimeout(() => this.exitGame(), 500)
    }
  }

  exitGame() {
    this.props.navigator.dismissModal({
      animationType: "none"
    })
  }

  render() {
    const {
      questionIndex,
      questionDone
    } = this.state

    const {
      questions
    } = this.props

    return (
      <ContainerView>
        <TopContainerView>
          <FlexedView>
            <Icon
              onPress={this.exitGame.bind(this)}
              name="exit-to-app"
              size={30}
              color={colors.lightGray} />
          </FlexedView>

          <ProgressBar
            completion={questionIndex / questions.length} />

          <FlexedView />
        </TopContainerView>

        <Question
          isInterlude={questionDone}
          questionDone={this.questionDone.bind(this)}
          question={questions[questionIndex]} />

        {
          questionDone &&
          <Button
            underlayColor={colors.green10l}
            onPress={this.nextQuestion.bind(this)}>
            <Text>
              CONTINUE
            </Text>
          </Button>
        }
      </ContainerView>
    )
  }
}

const FlexedView = styled.View`
  flex: 1;
`

const Button = styled.TouchableHighlight`
  width: 100px;
  height: 40px;
  align-items: center;
  justify-content: center;
  border-radius: 5px;
  position: absolute;
  right: 10px;
  bottom: 10px;
  background-color: ${colors.green};
`;

const Text = styled.Text`
  color: white;
  font-family: BrandonGrotesque-Bold;
`;

const ContainerView = styled.View`
  flex: 1;
  align-self: stretch;
`;

const TopContainerView = styled.View`
  flex: 1;
  align-self: stretch;
  flex-direction: row;
  align-items: center;
  justify-content: center;
  margin: 0px 5px;
`
