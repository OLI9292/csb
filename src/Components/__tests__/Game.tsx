import React from "react"
import renderer from "react-test-renderer"

import Game from "../Game"

it("renders correctly with defaults", () => {
  const game = renderer.create(<Game questions={[]} navigator={null} />).toJSON()
  expect(game).toMatchSnapshot()
})
