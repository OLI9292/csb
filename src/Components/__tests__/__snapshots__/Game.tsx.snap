// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with defaults 1`] = `
<View
  style={
    Array [
      Object {
        "alignSelf": "stretch",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "marginBottom": 0,
        "marginLeft": 20,
        "marginRight": 20,
        "marginTop": 20,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "alignSelf": "stretch",
          "flexBasis": 0,
          "flexDirection": "row",
          "flexGrow": 1,
          "flexShrink": 1,
          "justifyContent": "center",
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
          },
          undefined,
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={false}
        ellipsizeMode="tail"
        onPress={[Function]}
        style={
          Array [
            Object {
              "color": "#d9d9d9",
              "fontSize": 30,
            },
            undefined,
            Object {
              "fontFamily": "Material Design Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
          ]
        }
      >
        ïˆ†
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "flexBasis": 0,
            "flexGrow": 8,
            "flexShrink": 1,
            "justifyContent": "center",
            "marginBottom": 0,
            "marginLeft": 25,
            "marginRight": 25,
            "marginTop": 0,
          },
          undefined,
        ]
      }
    >
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "backgroundColor": "#d9d9d9",
              "borderRadius": 5,
              "height": 10,
              "width": "100%",
            },
            undefined,
          ]
        }
      />
      <View
        style={
          Object {
            "width": 0,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "backgroundColor": "#FFC31A",
                "borderRadius": 5,
                "height": 10,
                "marginTop": -10,
                "width": "100%",
              },
              undefined,
            ]
          }
        />
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
          },
          undefined,
        ]
      }
    />
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "flex-end",
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
        },
        undefined,
      ]
    }
  />
</View>
`;
